crates/bevy_app/src/app.rs

```diff
@@ -422,6 +425,48 @@ impl App {
         self
     }
 
+    /// Sets up a type implementing [`ComputedStates`].
+    ///
+    /// This method is idempotent: it has no effect when called again using the same generic type.
+    ///
+    /// For each source state the derived state depends on, it adds this state's derivation
+    /// to it's [`ComputeDependantStates<Source>`](bevy_ecs::schedule::ComputeDependantStates<S>) schedule.
+    pub fn add_computed_state<S: ComputedStates>(&mut self) -> &mut Self {
+        if !self
+            .world
+            .contains_resource::<Events<StateTransitionEvent<S>>>()
+        {
+            setup_state_transitions_in_world(&mut self.world);
+            self.add_event::<StateTransitionEvent<S>>();
+            let mut schedules = self.world.resource_mut::<Schedules>();
+            S::register_state_compute_systems_in_schedule(schedules.as_mut());
+        }
+
+        self
+    }
+
+    /// Sets up a type implementing [`SubStates`].
+    ///
+    /// This method is idempotent: it has no effect when called again using the same generic type.
+    ///
+    /// For each source state the derived state depends on, it adds this state's existance check
+    /// to it's [`ComputeDependantStates<Source>`](bevy_ecs::schedule::ComputeDependantStates<S>) schedule.
+    pub fn add_sub_state<S: SubStates>(&mut self) -> &mut Self {
+        if !self
+            .world
+            .contains_resource::<Events<StateTransitionEvent<S>>>()
+        {
+            setup_state_transitions_in_world(&mut self.world);
+            self.init_resource::<NextState<S>>();
```

Comment:
Okay, so substates can be changed and they should modify their parent state, few questions here:
1. How are transitions ran?
    Is the order always the same?
    - self -> parent -> other substates
    - parent -> self + other substates
1. What about NextState collisions?
    Parent and substate setting them at the same frame?
    Is there at least a constant order they get applied in?
1. What happens when you set a NextState to a substate that is currently disabled? (None)
    Will it get removed by next frame, will it be the initial state when this substate comes back to life?
    Will it force parent state to change into a variant that accommodates this substate?

Reply:
It may be worth noting some of the answers down in the function documentation, it's hard to dig through code when derive macros are involved

Reply:
For clarity - SubStates do not modify their parent state - they are independent states who's existence is managed by their parent states.

To answer the questions though:
1. substate transitions run similar to computed state transitions - `OnExit` of a substate is before the parent, and `OnTransition`/`OnEnter` is after the parent.
2. The parent state doesn't set the sub state, it is only used to determine if it exists. So unless the parent state change affects the existance of the SubState, there is no risk of collision. And a substate being added/removed takes priority over NextState for that SubState.
3. NextState gets ignored if the SubState doesn't exist.